pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeRangeSegmentPruner.java

```diff
@@ -0,0 +1,348 @@
+    onExternalViewChange(externalView, idealState, onlineSegments);
+  }
+
+  @Override
+  public void onExternalViewChange(ExternalView externalView, IdealState idealState, Set<String> onlineSegments) {
+    updateTimeRangeMSToSegmentSearchTree(onlineSegments);
+  }
+
+  private void updateTimeRangeMSToSegmentSearchTree(Set<String> onlineSegments) {
+    List<String> segments = new ArrayList<>(onlineSegments);
+    // atomic swap _timeRangeToSegmentMap for input online segments
+    List<String> segmentZKMetadataPaths = new ArrayList<>();
+    for (String segment: segments) {
+      segmentZKMetadataPaths.add(_segmentZKMetadataPathPrefix + segment);
+    }
+
+    List<ZNRecord> znRecords = _propertyStore.get(segmentZKMetadataPaths, null, AccessOption.PERSISTENT, true);
+
+    List<Long> startTimes = new ArrayList<>();
+    List<Long> endTimes = new ArrayList<>();
+
+    for (int i = 0; i < segments.size(); i++) {
+      String segment = segments.get(i);
+
+      long[] range = extractStartEndTimeMSFromSegmentZKMetaZNRecord(segment, znRecords.get(i));
+      startTimes.add(range[0]);
+      endTimes.add(range[1]);
+    }
+
+    _timeRangeMSToSegmentSearchTree = new IntervalST<String>(startTimes, endTimes, segments);
+  }
+
+
+  private long[] extractStartEndTimeMSFromSegmentZKMetaZNRecord(String segment, @Nullable ZNRecord znRecord) {
+    long[] range = {MIN_START_TIME, MAX_END_TIME};
+    // Segments without metadata or with invalid time range will be set with [min_start, max_end] and will not be pruned
+    if (znRecord == null) {
+      LOGGER.warn("Failed to find segment ZK metadata for segment: {}, table: {}", segment, _tableNameWithType);
+      return range;
+    }
+
+    long startTime = znRecord.getLongField(CommonConstants.Segment.START_TIME, -1);
+    long endTime = znRecord.getLongField(CommonConstants.Segment.END_TIME, -1);
+    if (startTime < 0 || endTime < 0 || startTime > endTime) {
+      LOGGER.warn("Failed to find valid end time for segment: {}, table: {}", segment, _tableNameWithType);
+      return range;
+    }
+
+    TimeUnit timeUnit = znRecord.getEnumField(CommonConstants.Segment.TIME_UNIT, TimeUnit.class, TimeUnit.DAYS);
+    range[0] = timeUnit.toMillis(startTime);
+    range[1] = timeUnit.toMillis(endTime);
+    return range;
+  }
+
+  @Override
+  public void refreshSegment(String segment) {
+    Set<String> segments;
+    if (_timeRangeMSToSegmentSearchTree == null) {
+      segments = new HashSet<>();
+    } else {
+      segments = _timeRangeMSToSegmentSearchTree.getAllValues();
+    }
+    if (!segment.contains(segment)) {
+      segments.add(segment);
+    }
+
+    updateTimeRangeMSToSegmentSearchTree(segments);
+  }
+
+  /**
+   * NOTE: Pruning is done by searching _timeRangeToSegmentSearchTree based on request time range and check if the results
+   *       are in the input segments. By doing so we will have run time O(M * logN) (N is the # of all online segments,
```

Comment:
The number of time ranges should also be part of Big O time complexity right?

Reply:
The worst case would be no segments share the same time range, so # time ranges would be the same as N in this case. Searching 1 intersected range will take O(logN) time, and get all results will take O(M*logN)